eclipse-collections-code-generator/src/main/resources/impl/collection/mutable/abstractMultiReaderPrimitiveCollection.stg

```diff
@@ -0,0 +1,554 @@
+    {
+        this.add(element);
+        return this;
+    }
+
+    public Mutable<name>Collection without(<type> element)
+    {
+        this.remove(element);
+        return this;
+    }
+
+    public Mutable<name>Collection withAll(<name>Iterable elements)
+    {
+        this.addAll(elements);
+        return this;
+    }
+
+    public Mutable<name>Collection withoutAll(<name>Iterable elements)
+    {
+        this.removeAll(elements);
+        return this;
+    }
+
+    public Mutable<name>Collection asUnmodifiable()
+    {
+        return new Unmodifiable<name>Collection(this);
+    }
+
+    public Mutable<name>Collection asSynchronized()
+    {
+        return this;
+    }
+
+    public Immutable<name>Collection toImmutable()
+    {
+        <readLock()>
+            return this.collection.toImmutable();
+        <readUnlock()>
+    }
+
+    public Lazy<name>Iterable asLazy()
+    {
+        return new Lazy<name>IterableAdapter(this);
+    }
+
+    public boolean contains(<type> value)
+    {
+        <readLock()>
+            return this.collection.contains(value);
+        <readUnlock()>
+    }
+
+    public boolean containsAll(<type>... source)
+    {
+        <readLock()>
+            return this.collection.containsAll(source);
+        <readUnlock()>
+    }
+
+    public boolean containsAll(<name>Iterable source)
+    {
+        <readLock()>
+            return this.collection.containsAll(source);
+        <readUnlock()>
+    }
+
+    public boolean add(<type> newItem)
+    {
+        <writeLock()>
+            return this.collection.add(newItem);
+        <writeUnlock()>
+    }
+
+    public boolean addAll(<type>... source)
+    {
+        <writeLock()>
+            return this.collection.addAll(source);
+        <writeUnlock()>
+    }
+
+    public boolean addAll(<name>Iterable source)
+    {
+        <writeLock()>
+            return this.collection.addAll(source);
+        <writeUnlock()>
+    }
+
+    public boolean remove(<type> value)
+    {
+        <writeLock()>
+            return this.collection.remove(value);
+        <writeUnlock()>
+    }
+
+    public boolean removeAll(<name>Iterable source)
+    {
+        <writeLock()>
+            return this.collection.removeAll(source);
+        <writeUnlock()>
+
+    }
+
+    public boolean removeAll(<type>... source)
+    {
+        <writeLock()>
+            return this.collection.removeAll(source);
+        <writeUnlock()>
+    }
+
+    public boolean retainAll(<name>Iterable source)
+    {
+        <writeLock()>
+            return this.collection.retainAll(source);
+        <writeUnlock()>
+    }
+
+    public boolean retainAll(<type>... source)
+    {
+        <writeLock()>
+            return this.collection.retainAll(source);
+        <writeUnlock()>
+    }
+
+    /**
+     * Makes copy of entire collection
+     */
+    public Mutable<name>Iterator <type>Iterator()
+    {
+        final <name>Iterator iter = toImmutable().<type>Iterator();
```

Comment:
The "object" collections do this, and I think we should consider doing the same here.

```
    @Override
    public Iterator<T> iterator()
    {
        throw new UnsupportedOperationException(
                "Iterator is not supported directly on MultiReader collections.  "
                        + "If you would like to use an iterator, you must either use withReadLockAndDelegate() or withWriteLockAndDelegate().");
    }
```